File: schema/runtime.go
package schema

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// ============================================================
// Load Mode
// ============================================================

// LoadMode controls how secure fields are handled.
type LoadMode string

const (
	// ReadMode uses secureJsonFields (boolean map) to check secret presence.
	ReadMode LoadMode = "read"

	// WriteMode uses secureJsonData (actual values) for full validation.
	WriteMode LoadMode = "write"
)

// ============================================================
// Input: Grafana datasource config
// ============================================================

// DatasourceConfig represents a Grafana datasource storage payload.
// This is an internal runtime representation — use NewDatasourceConfig
// to parse from a raw Grafana datasource resource.
//
// In Grafana's storage model, root-level fields (url, basicAuth, etc.)
// sit at the top level alongside jsonData and secureJsonData. The Root
// map groups those top-level fields for programmatic access.
type DatasourceConfig struct {
	// Root-level fields (url, basicAuth, basicAuthUser, etc.)
	// These are top-level in Grafana's storage, not nested under "root".
	Root map[string]any

	// Plugin-specific JSON config.
	JSONData map[string]any

	// Secret values (write path only).
	SecureJSONData map[string]any

	// Boolean map indicating which secrets are configured (read path).
	SecureJSONFields map[string]bool
}

// NewDatasourceConfig creates a DatasourceConfig from a flat Grafana
// datasource resource (as stored in the database or provisioning YAML).
// It extracts root fields, jsonData, secureJsonData, and secureJsonFields
// from the raw map.
func NewDatasourceConfig(raw map[string]any) DatasourceConfig {
	dc := DatasourceConfig{
		Root: make(map[string]any),
	}

	for k, v := range raw {
		switch k {
		case "jsonData":
			if m, ok := v.(map[string]any); ok {
				dc.JSONData = m
			}
		case "secureJsonData":
			if m, ok := v.(map[string]any); ok {
				dc.SecureJSONData = m
			}
		case "secureJsonFields":
			if m, ok := v.(map[string]any); ok {
				dc.SecureJSONFields = make(map[string]bool, len(m))
				for sk, sv := range m {
					if b, ok := sv.(bool); ok {
						dc.SecureJSONFields[sk] = b
					}
				}
			}
		default:
			dc.Root[k] = v
		}
	}

	return dc
}

// ============================================================
// Output: Load result
// ============================================================

// ConfigError represents a validation error for a specific field.
type ConfigError struct {
	// FieldID is the schema field ID (e.g. "auth.password").
	FieldID string `json:"fieldId"`

	// Path is the storage path (e.g. "secureJsonData.password").
	Path string `json:"path"`

	// Code is a machine-readable error code.
	Code string `json:"code"`

	// Message is a human-readable description.
	Message string `json:"message"`
}

// SecureState represents the state of a secure field.
type SecureState string

const (
	SecureUnset      SecureState = "unset"
	SecureConfigured SecureState = "configured"
	SecureUpdated    SecureState = "updated"
)

// FieldValue holds the extracted and resolved value for a single field.
type FieldValue struct {
	// Value is the extracted (and possibly defaulted) value.
	Value any

	// Source indicates where the value came from.
	Source ValueSource
}

// ValueSource indicates the origin of a field value.
type ValueSource string

const (
	SourceConfig  ValueSource = "config"  // extracted from config
	SourceDefault ValueSource = "default" // applied from schema default
	SourceNone    ValueSource = "none"    // not present, no default
)

// LoadResult is the output of LoadAndValidate.
type LoadResult struct {
	// Errors contains all validation errors found.
	Errors []ConfigError `json:"errors,omitempty"`

	// Values contains extracted field values keyed by field ID.
	Values map[string]FieldValue `json:"values,omitempty"`

	// SecureFields contains secret field states keyed by field ID.
	SecureFields map[string]SecureState `json:"secureFields,omitempty"`
}

// HasErrors returns true if any validation errors were found.
func (r *LoadResult) HasErrors() bool {
	return len(r.Errors) > 0
}

// ============================================================
// Pipeline: LoadAndValidate
// ============================================================

// LoadAndValidate extracts field values from a Grafana datasource config
// using the schema, applies defaults, validates values against field
// rules, and returns the result.
func LoadAndValidate(s *DatasourceConfigSchema, config DatasourceConfig, mode LoadMode) (*LoadResult, error) {
	if err := s.Validate(); err != nil {
		return nil, fmt.Errorf("invalid schema: %w", err)
	}

	result := &LoadResult{
		Values:       make(map[string]FieldValue),
		SecureFields: make(map[string]SecureState),
	}

	for i := range s.Fields {
		loadField(&s.Fields[i], config, mode, result)
	}

	return result, nil
}

// loadField processes a single top-level field.
func loadField(f *ConfigField, config DatasourceConfig, mode LoadMode, result *LoadResult) {
	if f.Kind == VirtualField {
		return
	}

	// Secure field handling
	if f.Target != nil && *f.Target == SecureJSONTarget {
		loadSecureField(f, config, mode, result)
		return
	}

	// Extract value
	raw := extractValue(config, f)

	// Expand indexedPair storage mapping
	if f.Storage != nil && f.Storage.Type == IndexedPairMapping {
		raw = expandIndexedPair(config, f)
	}

	// Apply default
	source := SourceConfig
	if raw == nil {
		if f.DefaultValue != nil {
			raw = f.DefaultValue
			source = SourceDefault
		} else {
			source = SourceNone
		}
	}

	result.Values[f.ID] = FieldValue{Value: raw, Source: source}

	// Validate
	validateFieldValue(f, raw, result)

	// Recurse into array item fields for validation
	if f.ValueType == ArrayType && f.Item != nil && raw != nil {
		if arr, ok := raw.([]any); ok {
			validateArrayItems(f, arr, result)
		}
	}
}

// loadSecureField handles fields targeting secureJsonData.
func loadSecureField(f *ConfigField, config DatasourceConfig, mode LoadMode, result *LoadResult) {
	switch mode {
	case WriteMode:
		val := config.SecureJSONData[f.Key]
		if val != nil {
			result.SecureFields[f.ID] = SecureUpdated
			result.Values[f.ID] = FieldValue{Value: val, Source: SourceConfig}
		} else {
			result.SecureFields[f.ID] = SecureUnset
			result.Values[f.ID] = FieldValue{Value: nil, Source: SourceNone}
		}
		// Validate required in write mode
		if f.Required && val == nil {
			result.Errors = append(result.Errors, ConfigError{
				FieldID: f.ID,
				Path:    f.Path(),
				Code:    "required",
				Message: fmt.Sprintf("field %s is required", f.ID),
			})
		}

	case ReadMode:
		if config.SecureJSONFields[f.Key] {
			result.SecureFields[f.ID] = SecureConfigured
		} else {
			result.SecureFields[f.ID] = SecureUnset
			if f.Required {
				result.Errors = append(result.Errors, ConfigError{
					FieldID: f.ID,
					Path:    f.Path(),
					Code:    "required",
					Message: fmt.Sprintf("field %s is required", f.ID),
				})
			}
		}
		// No value extraction in read mode for secrets
		result.Values[f.ID] = FieldValue{Value: nil, Source: SourceNone}
	}
}

// ============================================================
// Value extraction
// ============================================================

// extractValue reads a field's value from the correct config location.
func extractValue(config DatasourceConfig, f *ConfigField) any {
	if f.Target == nil {
		return nil
	}

	switch *f.Target {
	case RootTarget:
		return config.Root[f.Key]

	case JSONDataTarget:
		if config.JSONData == nil {
			return nil
		}
		if f.Section == "" {
			return config.JSONData[f.Key]
		}
		return extractFromSection(config.JSONData, f.Section, f.Key)

	case SecureJSONTarget:
		// Handled separately by loadSecureField
		return nil

	default:
		return nil
	}
}

// extractFromSection walks a dotted section path through a nested map
// and returns the value at the final key.
func extractFromSection(data map[string]any, section, key string) any {
	segments := strings.Split(section, ".")
	current := data

	for _, seg := range segments {
		v, ok := current[seg]
		if !ok {
			return nil
		}
		m, ok := v.(map[string]any)
		if !ok {
			return nil
		}
		current = m
	}

	return current[key]
}

// ============================================================
// IndexedPair expansion
// ============================================================

// expandIndexedPair reads indexed key/value pairs from the config
// and assembles them into an array of objects.
func expandIndexedPair(config DatasourceConfig, f *ConfigField) any {
	if f.Storage == nil || f.Storage.Type != IndexedPairMapping {
		return nil
	}
	if f.Storage.Key == nil || f.Storage.Value == nil {
		return nil
	}
	if f.Item == nil || len(f.Item.Fields) < 2 {
		return nil
	}

	startIndex := 1
	if f.Storage.StartIndex != nil {
		startIndex = *f.Storage.StartIndex
	}

	keyPattern := f.Storage.Key.Pattern
	valuePattern := f.Storage.Value.Pattern

	// Determine item field keys (first field = key name, second = value name)
	keyFieldKey := f.Item.Fields[0].Key
	valueFieldKey := f.Item.Fields[1].Key

	var items []any

	for i := startIndex; ; i++ {
		idx := strconv.Itoa(i)
		kName := strings.ReplaceAll(keyPattern, "{index}", idx)
		vName := strings.ReplaceAll(valuePattern, "{index}", idx)

		kVal := getFromTarget(config, f.Storage.Key.Target, kName)
		if kVal == nil {
			break // Stop when we hit a gap
		}

		vVal := getFromTarget(config, f.Storage.Value.Target, vName)

		item := map[string]any{
			keyFieldKey:   kVal,
			valueFieldKey: vVal,
		}
		items = append(items, item)
	}

	if items == nil {
		return nil
	}
	return items
}

// getFromTarget reads a key from the appropriate config section.
func getFromTarget(config DatasourceConfig, target TargetLocation, key string) any {
	switch target {
	case RootTarget:
		return config.Root[key]
	case JSONDataTarget:
		if config.JSONData == nil {
			return nil
		}
		return config.JSONData[key]
	case SecureJSONTarget:
		if config.SecureJSONData == nil {
			return nil
		}
		return config.SecureJSONData[key]
	default:
		return nil
	}
}

// ============================================================
// Value validation
// ============================================================

// validateFieldValue validates an extracted value against field rules.
func validateFieldValue(f *ConfigField, value any, result *LoadResult) {
	// Required check
	if f.Required && value == nil {
		result.Errors = append(result.Errors, ConfigError{
			FieldID: f.ID,
			Path:    f.Path(),
			Code:    "required",
			Message: fmt.Sprintf("field %s is required", f.ID),
		})
		return // No further validation on nil
	}

	if value == nil {
		return
	}

	// Type check
	if err := checkValueType(value, f.ValueType); err != nil {
		result.Errors = append(result.Errors, ConfigError{
			FieldID: f.ID,
			Path:    f.Path(),
			Code:    "type_mismatch",
			Message: fmt.Sprintf("field %s: %s", f.ID, err),
		})
		return // No further validation on wrong type
	}

	// Validation rules
	for _, rule := range f.Validations {
		if err := evaluateRule(value, rule, f); err != nil {
			msg := err.Error()
			if rule.Message != "" {
				msg = rule.Message
			}
			result.Errors = append(result.Errors, ConfigError{
				FieldID: f.ID,
				Path:    f.Path(),
				Code:    string(rule.Type),
				Message: msg,
			})
		}
	}
}

// checkValueType verifies that a value matches the expected valueType.
func checkValueType(value any, expected ValueType) error {
	switch expected {
	case StringType:
		if _, ok := value.(string); !ok {
			return fmt.Errorf("expected string, got %T", value)
		}
	case NumberType:
		switch value.(type) {
		case float64, float32, int, int64, int32:
			// ok — JSON numbers unmarshal as float64
		default:
			return fmt.Errorf("expected number, got %T", value)
		}
	case BooleanType:
		if _, ok := value.(bool); !ok {
			return fmt.Errorf("expected boolean, got %T", value)
		}
	case ArrayType:
		if _, ok := value.([]any); !ok {
			return fmt.Errorf("expected array, got %T", value)
		}
	case ObjectType:
		if _, ok := value.(map[string]any); !ok {
			return fmt.Errorf("expected object, got %T", value)
		}
	}
	return nil
}

// evaluateRule validates a value against a single validation rule.
func evaluateRule(value any, rule FieldValidationRule, f *ConfigField) error {
	switch rule.Type {
	case PatternValidation:
		s, ok := value.(string)
		if !ok {
			return nil // Type mismatch caught separately
		}
		matched, err := regexp.MatchString(rule.Pattern, s)
		if err != nil {
			return fmt.Errorf("invalid regex pattern %q: %w", rule.Pattern, err)
		}
		if !matched {
			return fmt.Errorf("value does not match pattern %q", rule.Pattern)
		}

	case RangeValidation:
		n, ok := toFloat64(value)
		if !ok {
			return nil
		}
		if rule.Min != nil && n < *rule.Min {
			return fmt.Errorf("value %v is below minimum %v", n, *rule.Min)
		}
		if rule.Max != nil && n > *rule.Max {
			return fmt.Errorf("value %v exceeds maximum %v", n, *rule.Max)
		}

	case LengthValidation:
		s, ok := value.(string)
		if !ok {
			return nil
		}
		l := float64(len(s))
		if rule.Min != nil && l < *rule.Min {
			return fmt.Errorf("length %d is below minimum %v", len(s), *rule.Min)
		}
		if rule.Max != nil && l > *rule.Max {
			return fmt.Errorf("length %d exceeds maximum %v", len(s), *rule.Max)
		}

	case ItemCountValidation:
		arr, ok := value.([]any)
		if !ok {
			return nil
		}
		l := float64(len(arr))
		if rule.Min != nil && l < *rule.Min {
			return fmt.Errorf("item count %d is below minimum %v", len(arr), *rule.Min)
		}
		if rule.Max != nil && l > *rule.Max {
			return fmt.Errorf("item count %d exceeds maximum %v", len(arr), *rule.Max)
		}

	case AllowedValuesValidation:
		for _, allowed := range rule.Values {
			if fmt.Sprintf("%v", value) == fmt.Sprintf("%v", allowed) {
				return nil
			}
		}
		return fmt.Errorf("value %v is not in allowed values %v", value, rule.Values)

	case CustomValidation:
		// CEL not evaluated — skip silently
		return nil
	}

	return nil
}

// validateArrayItems validates each element of an array against item field rules.
func validateArrayItems(f *ConfigField, arr []any, result *LoadResult) {
	if f.Item == nil {
		return
	}

	for i, elem := range arr {
		if f.Item.ValueType == ObjectType && len(f.Item.Fields) > 0 {
			obj, ok := elem.(map[string]any)
			if !ok {
				result.Errors = append(result.Errors, ConfigError{
					FieldID: f.ID,
					Path:    fmt.Sprintf("%s[%d]", f.Path(), i),
					Code:    "type_mismatch",
					Message: fmt.Sprintf("field %s[%d]: expected object, got %T", f.ID, i, elem),
				})
				continue
			}
			for j := range f.Item.Fields {
				sub := &f.Item.Fields[j]
				val := obj[sub.Key]
				if val == nil && sub.DefaultValue != nil {
					val = sub.DefaultValue
				}
				validateFieldValue(sub, val, result)
			}
		}
	}
}

// toFloat64 converts numeric types to float64 for comparison.
func toFloat64(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case int32:
		return float64(n), true
	default:
		return 0, false
	}
}

File: schema/runtime.ts
import type {
    DatasourceConfigSchema,
    ConfigField,
    FieldValidationRule,
    TargetLocation,
    ValueType,
} from "./schema"

// ============================================================
// Load Mode
// ============================================================

export type LoadMode = "read" | "write"

// ============================================================
// Input: Grafana datasource config
// ============================================================

export interface DatasourceConfig {
    root: Record<string, unknown>
    jsonData?: Record<string, unknown>
    secureJsonData?: Record<string, unknown>
    secureJsonFields?: Record<string, boolean>
}

/** Parse a flat Grafana datasource resource into DatasourceConfig. */
export function newDatasourceConfig(raw: Record<string, unknown>): DatasourceConfig {
    const dc: DatasourceConfig = { root: {} }

    for (const [k, v] of Object.entries(raw)) {
        switch (k) {
            case "jsonData":
                if (v && typeof v === "object" && !Array.isArray(v)) {
                    dc.jsonData = v as Record<string, unknown>
                }
                break
            case "secureJsonData":
                if (v && typeof v === "object" && !Array.isArray(v)) {
                    dc.secureJsonData = v as Record<string, unknown>
                }
                break
            case "secureJsonFields":
                if (v && typeof v === "object" && !Array.isArray(v)) {
                    dc.secureJsonFields = v as Record<string, boolean>
                }
                break
            default:
                dc.root[k] = v
        }
    }

    return dc
}

// ============================================================
// Output
// ============================================================

export interface ConfigError {
    fieldId: string
    path: string
    code: string
    message: string
}

export type SecureState = "unset" | "configured" | "updated"

export type ValueSource = "config" | "default" | "none"

export interface FieldValue {
    value: unknown
    source: ValueSource
}

export interface LoadResult {
    errors: ConfigError[]
    values: Record<string, FieldValue>
    secureFields: Record<string, SecureState>
}

// ============================================================
// Pipeline
// ============================================================

export function loadAndValidate(
    schema: DatasourceConfigSchema,
    config: DatasourceConfig,
    mode: LoadMode,
): LoadResult {
    const result: LoadResult = {
        errors: [],
        values: {},
        secureFields: {},
    }

    for (const field of schema.fields) {
        loadField(field, config, mode, result)
    }

    return result
}

function loadField(
    f: ConfigField,
    config: DatasourceConfig,
    mode: LoadMode,
    result: LoadResult,
): void {
    if (f.kind === "virtual") {
        return
    }

    if (f.target === "secureJsonData") {
        loadSecureField(f, config, mode, result)
        return
    }

    let raw = extractValue(config, f)

    if (f.storage?.type === "indexedPair") {
        raw = expandIndexedPair(config, f)
    }

    let source: ValueSource = "config"
    if (raw == null) {
        if (f.defaultValue !== undefined) {
            raw = f.defaultValue
            source = "default"
        } else {
            source = "none"
        }
    }

    result.values[f.id] = { value: raw ?? null, source }

    validateFieldValue(f, raw ?? null, result)

    if (f.valueType === "array" && f.item && Array.isArray(raw)) {
        validateArrayItems(f, raw, result)
    }
}

function loadSecureField(
    f: ConfigField,
    config: DatasourceConfig,
    mode: LoadMode,
    result: LoadResult,
): void {
    const path = fieldPath(f)

    if (mode === "write") {
        const val = config.secureJsonData?.[f.key]
        if (val != null) {
            result.secureFields[f.id] = "updated"
            result.values[f.id] = { value: val, source: "config" }
        } else {
            result.secureFields[f.id] = "unset"
            result.values[f.id] = { value: null, source: "none" }
        }
        if (f.required && val == null) {
            result.errors.push({
                fieldId: f.id,
                path,
                code: "required",
                message: `field ${f.id} is required`,
            })
        }
    } else {
        if (config.secureJsonFields?.[f.key]) {
            result.secureFields[f.id] = "configured"
        } else {
            result.secureFields[f.id] = "unset"
            if (f.required) {
                result.errors.push({
                    fieldId: f.id,
                    path,
                    code: "required",
                    message: `field ${f.id} is required`,
                })
            }
        }
        result.values[f.id] = { value: null, source: "none" }
    }
}

// ============================================================
// Value extraction
// ============================================================

function extractValue(config: DatasourceConfig, f: ConfigField): unknown {
    if (!f.target) return undefined

    switch (f.target) {
        case "root":
            return config.root[f.key]
        case "jsonData": {
            if (!config.jsonData) return undefined
            if (!f.section) return config.jsonData[f.key]
            return extractFromSection(config.jsonData, f.section, f.key)
        }
        case "secureJsonData":
            return undefined // handled by loadSecureField
        default:
            return undefined
    }
}

function extractFromSection(
    data: Record<string, unknown>,
    section: string,
    key: string,
): unknown {
    const segments = section.split(".")
    let current: Record<string, unknown> = data

    for (const seg of segments) {
        const v = current[seg]
        if (v == null || typeof v !== "object" || Array.isArray(v)) {
            return undefined
        }
        current = v as Record<string, unknown>
    }

    return current[key]
}

// ============================================================
// IndexedPair expansion
// ============================================================

function expandIndexedPair(config: DatasourceConfig, f: ConfigField): unknown[] | undefined {
    const storage = f.storage
    if (!storage || storage.type !== "indexedPair") return undefined
    if (!("key" in storage) || !("value" in storage)) return undefined
    if (!f.item?.fields || f.item.fields.length < 2) return undefined

    const startIndex = ("startIndex" in storage && storage.startIndex != null)
        ? storage.startIndex
        : 1

    const keyPattern = storage.key.pattern
    const valuePattern = storage.value.pattern
    const keyFieldKey = f.item.fields[0].key
    const valueFieldKey = f.item.fields[1].key

    const items: unknown[] = []

    for (let i = startIndex; ; i++) {
        const idx = String(i)
        const kName = keyPattern.replace("{index}", idx)
        const vName = valuePattern.replace("{index}", idx)

        const kVal = getFromTarget(config, storage.key.target, kName)
        if (kVal == null) break

        const vVal = getFromTarget(config, storage.value.target, vName)

        items.push({
            [keyFieldKey]: kVal,
            [valueFieldKey]: vVal,
        })
    }

    return items.length > 0 ? items : undefined
}

function getFromTarget(
    config: DatasourceConfig,
    target: TargetLocation,
    key: string,
): unknown {
    switch (target) {
        case "root":
            return config.root[key]
        case "jsonData":
            return config.jsonData?.[key]
        case "secureJsonData":
            return config.secureJsonData?.[key]
        default:
            return undefined
    }
}

// ============================================================
// Value validation
// ============================================================

function validateFieldValue(
    f: ConfigField,
    value: unknown,
    result: LoadResult,
): void {
    const path = fieldPath(f)

    if (f.required && value == null) {
        result.errors.push({
            fieldId: f.id,
            path,
            code: "required",
            message: `field ${f.id} is required`,
        })
        return
    }

    if (value == null) return

    const typeErr = checkValueType(value, f.valueType)
    if (typeErr) {
        result.errors.push({
            fieldId: f.id,
            path,
            code: "type_mismatch",
            message: `field ${f.id}: ${typeErr}`,
        })
        return
    }

    if (f.validations) {
        for (const rule of f.validations) {
            const err = evaluateRule(value, rule)
            if (err) {
                result.errors.push({
                    fieldId: f.id,
                    path,
                    code: rule.type,
                    message: rule.message ?? err,
                })
            }
        }
    }
}

function checkValueType(value: unknown, expected: ValueType): string | null {
    switch (expected) {
        case "string":
            return typeof value === "string" ? null : `expected string, got ${typeof value}`
        case "number":
            return typeof value === "number" ? null : `expected number, got ${typeof value}`
        case "boolean":
            return typeof value === "boolean" ? null : `expected boolean, got ${typeof value}`
        case "array":
            return Array.isArray(value) ? null : `expected array, got ${typeof value}`
        case "object":
            return (value !== null && typeof value === "object" && !Array.isArray(value))
                ? null
                : `expected object, got ${typeof value}`
        default:
            return null
    }
}

function evaluateRule(value: unknown, rule: FieldValidationRule): string | null {
    switch (rule.type) {
        case "pattern": {
            if (typeof value !== "string") return null
            const re = new RegExp(rule.pattern)
            return re.test(value) ? null : `value does not match pattern "${rule.pattern}"`
        }

        case "range": {
            if (typeof value !== "number") return null
            if (rule.min != null && value < rule.min) {
                return `value ${value} is below minimum ${rule.min}`
            }
            if (rule.max != null && value > rule.max) {
                return `value ${value} exceeds maximum ${rule.max}`
            }
            return null
        }

        case "length": {
            if (typeof value !== "string") return null
            if (rule.min != null && value.length < rule.min) {
                return `length ${value.length} is below minimum ${rule.min}`
            }
            if (rule.max != null && value.length > rule.max) {
                return `length ${value.length} exceeds maximum ${rule.max}`
            }
            return null
        }

        case "itemCount": {
            if (!Array.isArray(value)) return null
            if (rule.min != null && value.length < rule.min) {
                return `item count ${value.length} is below minimum ${rule.min}`
            }
            if (rule.max != null && value.length > rule.max) {
                return `item count ${value.length} exceeds maximum ${rule.max}`
            }
            return null
        }

        case "allowedValues": {
            const vals = rule.values ?? []
            for (const allowed of vals) {
                if (String(value) === String(allowed)) return null
            }
            return `value ${value} is not in allowed values [${vals.join(", ")}]`
        }

        case "custom":
            return null // CEL not evaluated
    }

    return null
}

function validateArrayItems(
    f: ConfigField,
    arr: unknown[],
    result: LoadResult,
): void {
    if (!f.item) return

    for (let i = 0; i < arr.length; i++) {
        const elem = arr[i]
        if (f.item.valueType === "object" && f.item.fields && f.item.fields.length > 0) {
            if (typeof elem !== "object" || elem === null || Array.isArray(elem)) {
                result.errors.push({
                    fieldId: f.id,
                    path: `${fieldPath(f)}[${i}]`,
                    code: "type_mismatch",
                    message: `field ${f.id}[${i}]: expected object, got ${typeof elem}`,
                })
                continue
            }
            const obj = elem as Record<string, unknown>
            for (const sub of f.item.fields) {
                let val = obj[sub.key]
                if (val == null && sub.defaultValue !== undefined) {
                    val = sub.defaultValue
                }
                validateFieldValue(sub, val ?? null, result)
            }
        }
    }
}

// ============================================================
// Helpers
// ============================================================

function fieldPath(f: ConfigField): string {
    if (!f.target) return f.key
    if (f.section) return `${f.target}.${f.section}.${f.key}`
    return `${f.target}.${f.key}`
}
